src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotUseReferenceEqualsWithValueTypes.cs

```diff
@@ -0,0 +1,98 @@
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.DoNotUseReferenceEqualsWithValueTypesMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.DoNotUseReferenceEqualsWithValueTypesDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.MicrosoftCodeAnalysisCorrectness,
+            RuleLevel.BuildWarning,
+            description: s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                var objectType = compilationStartContext.Compilation.GetSpecialType(SpecialType.System_Object);
+
+                // Without these symbols the rule cannot run
+                if (objectType == null)
+                {
+                    return;
+                }
+
+                var referenceEqualsMethodGroup = objectType.GetMembers("ReferenceEquals").OfType<IMethodSymbol>();
+                var referenceEqualsMethod = referenceEqualsMethodGroup.GetFirstOrDefaultMemberWithParameterInfos(
+                    GetParameterInfo(objectType),
+                    GetParameterInfo(objectType));
+
+                if (referenceEqualsMethod == null)
+                {
+                    return;
+                }
+
+                compilationStartContext.RegisterOperationAction(operationContext =>
+                {
+                    var invocationExpression = (IInvocationOperation)operationContext.Operation;
+                    var targetMethod = invocationExpression.TargetMethod;
+
+                    if (!referenceEqualsMethod.Equals(targetMethod))
+                    {
+                        return;
+                    }
+
+                    foreach (var argument in invocationExpression.Arguments)
+                    {
+                        var val = argument.Value;
+
+                        if (val is IConversionOperation conversion)
```

Comment:
To handle multiple conversions - for example an implicit conversion to a value type on top of a user defined conversion.
```suggestion
                        val = val.WalkDownConversion();
```

Reply:
Really I think the question here is just "is the last step a boxing conversion". `ReferenceEquals((SomeCacheType)5, other)` is sensible if SomeCacheType is a reference type and the conversion operator was doing a dictionary lookup by the integer ID; so if WalkDownConversion is going to go all the way to the `5` then it seems like the wrong approach.

Reply:
Sounds good!